core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -864,6 +778,111 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  /*
+   * Note: This method can be used as a callback in a different request thread. Ensure that correct RequestLocal
+   * is passed when executing this method. Accessing non-thread-safe data structures should be avoided if possible.
+   */
+  private def appendEntries(allEntries: Map[TopicPartition, MemoryRecords],
+                            internalTopicsAllowed: Boolean,
+                            origin: AppendOrigin,
+                            requiredAcks: Short,
+                            verificationGuards: Map[TopicPartition, VerificationGuard],
```

Comment:
1. Are we sure that the Maps being passed here are not concurrently modified by two threads? Asking because they are not thread safe (they aren't concurrent hash maps). I haven't looked in details about the params passed here so please feel free to say that you have already verified this.

2. How can we prevent future bugs where someone mutates one of these parameters in perhaps other parts of the code such as Kafka APIs where appendEntries is called from without knowing that all these data structures are supposed to be thread safe?

One option to prevent future bugs could be to pass a deep copy of these objects to the callback (instead of a copy of the reference). Another thing we can do is to restrict the number of parameters required by the callback.

Having said that, have we already considered not having this as a callback? I understand that it has drawbacks since it is in critical produce path but maybe that would be acceptable given the complexity of ensuring thread safety here?

Reply:
These maps are not mutable right?

Reply:
> These maps are not mutable right?

Don't know. I haven't really checked. I wanted to ask you instead if you have verified that :)

Reply:
Sorry for my phrasing. I was saying they are mutable. 😅

Reply:
> Sorry for my phrasing. I was saying they are mutable. 😅

Did you mean to say immutable here? If they are mutable, then we need to ensure that they are not concurrently mutated in different threads.

Reply:
Yes. I'm having trouble speaking today. Map in scala is immutable. 🤦‍♀️ Apologies for confusion.

Reply:
Actually, I don't think this method enforces that these maps are immutable. That said, my understanding is that we basically suspend processing and hence there is only one thread accessing these structures at a time. Is that correct? If so, the main thing we need to do is ensure we safely publish these data structures, they don't have to be immutable necessarily.

Reply:
That's correct -- we don't create new concurrent access, i.e. this callback is not called on multiple threads at the same time, all we do here is potentially executing this callback on a different thread.  Actually I think "Accessing non-thread-safe data structures should be avoided if possible" comment is misleading here, the only thing that we have to avoid is capturing thread local (or effectively thread local) objects in the callback.  Other objects can be as thread safe as they would be if this method was called directly on this thread (i.e. if an object was rooted on the call stack it doesn't have to be thread safe because it's not going to be used concurrently from multiple threads).

BTW, we have plenty of cases in Kafka (both client and server) where we pass a callback to continue execution of a single-threaded logical task after an RPC call is complete (most likely on a different thread), so I don't think we're dealing with unique pattern here.  The only thing that is unique (and tricky) here is the RequestLocal that disguises as a context of a logical task (arguments are generally are logical task context), but it is actually a concrete physical thread context that would be different (or missing completely, say if we executed this callback on the network thread) if we continued execution of this logical task on a different thread.

I've already pointed out elsewhere that (in a future change) we should probably remove the RequestLocal from the argument list and make it a thread-local (which it effectively is) and just use it directly where it's needed.

Reply:
A lot of callback code runs on the same thread though - the cases where a different thread is used require more care and have caused a bunch of pain in the past.

Reply:
I think how much work is done in the callback is more a quantitative difference rather than a qualitative difference -- how much work needs to be done to treat it as a new pattern that requires some special care (and what is that care, can we qualify)?    I agree that if the functionality in the callback just happened to be after the point of use of the RequestLocal then we wouldn't hit the problem (even if we captured it in the callback), but it seems to a matter of just random "luck".  I don't think "don't be unlucky" is a useful engineering guidance.  On the other hand, saying "don't pass thread local as an argument, access it at the point of use as a thread local instead" would avoid these cases.

But maybe the first point to discuss is do we agree that the root cause is the fact that RequestLocal is bound to the executing thread context (effectively thread local passed as an argument) and it's not a typical pattern in Java or do we think that it's something else?  At some point I think there was a theory that the machinery for re-scheduling the callback on the request thread pool could be a culprit, but I think we all agree, that the problem would be the same even if the callback was executed directly on the network thread (or pretty much in any other way other than trying to schedule it on the exact same thread, which is doable, btw, but would hinder performance).

Reply:
I think you misunderstood my point. I was saying that there are two types of callbacks: (1) execute on the same thread (2) execute on a different thread. They're not the same and the latter require more care.

Reply:
Thanks for the discussion here folks. My point about the immutability of the map is that we can't put new objects in the map and the values appeared to be final. I agree that the method itself (and the argument for wrap) doesn't guarantee immutability and/or thread safety. 

I think we can agree the the RequestLocal object provides specific challenges for callbacks that require it. Specifically callbacks that may execute on a different thread may see issues. 

How do we want to move forward from here? I can modify my comment and the class to remove the comments about thread safety etc and focus on the fact that the method may execute on a different thread. (ie name the class something that reflects that) Were there any other changes we think are necessary?

Reply:
I agree, it's just not exactly clear what an actionable practical recommendation could be derived from that.

Reply:
(Race condition -- my reply was to Ismael w.r.t. "more care required").

Reply:
1. On the server side, some of the callbacks (Purgatory, the one introduced in KAFKA-14561, etc) are called in a different thread than the caller. I am not sure if it's possible/desirable to change all of those callbacks to be run on the same thread. Given that model, I agree that the wide exposure of ThreadLocal seems potentially dangerous. I pinged https://github.com/apache/kafka/pull/9220 to see if we could consider an alternative approach. It might be easier to fix the `ThreadLocal` thing first before fixing this issue.
2. In general, callbacks on different threads are tricky to get right. For example, we spend a lot of time to fix the deadlock issues related to Purgatory (https://issues.apache.org/jira/browse/KAFKA-8334). For this newly introduced callback, I am wondering about the locking model. We have the following path that's called under the GroupMetadata lock. `GroupCoordinator.doSyncGroup (hold GroupMetadata lock) => groupManager.storeGroup => appendForGroup => replicaManager.appendRecords => replicaManager.appendEntries`. However, if we register the callback, `replicaManager.appendEntries` will be called without holding the `GroupMetadata` lock. Is that safe?

Reply:
>  actionable practical recommendation

I can't think of an idea other than passing deep copy of arguments. I have also considered suggesting wrapping in Collections.unmodifiable or Collections.synchronizedCollection but none of these ideas are good enough. 

We need to re-think callback invocation pattern in Kafka and see if we can avoid this pattern of passing reference of in-memory state around threads. But I am happy to keep it our of this PR and discuss that in scope of JIRA that Ismael created.

Reply:
>  I am wondering about the locking model. We have the following path that's called under the GroupMetadata lock. GroupCoordinator.doSyncGroup (hold GroupMetadata lock) => groupManager.storeGroup => appendForGroup => replicaManager.appendRecords => replicaManager.appendEntries. However, if we register the callback, replicaManager.appendEntries will be called without holding the GroupMetadata lock. Is that safe?

Hmm. Do we need the lock to write to log or just to updaet the in-memory state. I can take another look at this path.

Reply:
It seems that we do have the lock for the in-memory state. It would just be the append itself that is not under the lock.

Reply:
Not sure how strongly we need the `GroupMetadata` lock while appending to the log. But I think the intention is to make sure that all writes to the log for a group are serialized. Updating the in-memory state typically happens in the callback when the HWM advances and is also protected by the `GroupMetadata` lock.

Reply:
> It would just be the append itself that is not under the lock.

Hmm, is that true? We have the following code.

```
GroupCoordinator.doSyncGroup
    group.inLock {
...
          groupManager.storeGroup
... 
    }
```

Reply:
I am saying that before the append was under the lock. With the verification change, now it is not. I think we are saying the same thing here. We need to figure out if the append not under the lock is safe.

Reply:
It seems that storing the group coordinator state is not transactional, so verification code path shouldn't be used.  But I agree with question of the locking model in the group coordinator and whether it'd work correctly with more asynchronous approach.

Reply:
There is a verification state for the doCommitTxnOffset path which is mostly the same path (execept for using storeOffsets instead of storeGroup)

Reply:
Hmm, without the lock, an offset commit may be reordered after a group state change. To make it safe, we would need to reacquire the group lock after validation completes so that we could check again whether the request should be accepted.

Reply:
@hachikuji : I am wondering the reason for the group coordinator to hold a group lock while appending to the log. It seems it's just to make sure the group state checking/updating and the appending to the log can be done atomically. I am not sure if there is any ordering guarantee since requests from different clients could come in arbitrary order. Is that the case?

Reply:
I think it's more a concern about group changes that are made while the verification is ongoing. Offline we were discussing what would happen if we bumped an epoch/rebalance for example

Reply:
@junrao If the lock cannot protect the write, then the validation in the group coordinator cannot be relied on. In principle, it would be possible for an older offset commit to override a newer one for example. A more likely scenario might be for a group to rebalance while an offset commit is awaiting validation. This would allow a fenced consumer to be able to commit offsets. Without the lock, I think we'd have to introduce additional bookkeeping to track dependencies between inflight writes.

Reply:
Given this is separate to the requestLocal issue, does it make sense to file a JIRA and fix there? We can still consider it for a patch release if we think this is a regression.

Reply:
Sorry, I am late here. I actually faced a similar issue while implementing this API in the new group coordinator. The scenario is more or less the same. We need to guarantee that the update to the state machine and to the log are done together to guarantee the ordering of the changes.

Given this, I wonder if we should just change how we do it. I was about to re-implement this part to work as follow:
1) Verify the transaction and possibly go the transaction coordinator;
2) Execute the actual transactional offset commit logic (update state + write).

I think that this is easier than introducing extra bookkeeping to track dependencies. What do you think? Note that I need to change it anyway for the new coordinator.

One way would be to bring this logic to the KafkaApis layer or we could rework the write logic within the group coordinator. Would it make sense?

Reply:
That makes sense to me. So maybe the verification would go through a different flow where the callback would include both the state update and the write via appendEntries. The main question I have is whether we should consider this as a separate JIRA and do in a follow-up PR. I am inclined to go with this option. 
(But do it ASAP for 3.6.1).

Reply:
@dajac Yes, that is similar to what I had in mind. We need to reacquire the lock after validation in order to verify whether the write should still be accepted. And we still need to protect the write itself. Your approach seems like a simpler implementation of this. The only downside I can think of is that we may add the the partition to the transaction but still ultimately discard the write, but this is probably not a common case.

Reply:
I guess that proposal is that when serving an `OffsetCommit` request, KafkaAPIs first determines the partitions for txn validation. It then initiates the txn validation asychnronously. Once the validation is done, it calls `ReplicaManager.append` w/o further txn validation.

If we do this, I am wondering if we should just do the same for regular produce requests too. This way, we don't need the callback logic for txn validation in `ReplicaManager.append`.

Reply:
This is effectively what we are doing right? appendEntries is just replica manager append w/o further txn validation.

Reply:
Hmm, I think it would break the abstraction of the appendRecords being a fundamentally asynchronous call, i.e. it has the syntax of being asynchronous (i.e. completion callback is surely called in a different thread) but still a concrete implementation expects that a specific part of the call is synchronous, but the code has no indication which one (maybe some day we'd use async IO and again make it more asynchronous then the code around it would expect).  Like why would the group coordinator need to know about transaction verification flow which has it's own trickiness with verification guard and may actually (likely) be different for KIP-890 part2?

I think it would be easier for separation of concerns if the lock is held only around im-memory  state updates and just transition the state into a "pending state", so that the new updated would be queued or failed (with a retriable error) until async call is done.

Then we wouldn't have a situation where we have a lock with a syntax that looks like "lock around in-memory update + call", but with the desired semantics "[lock] [in-memory update] [some part of call] [unlock] [some other part of call]" and the place of [unlock] is fairly deep in logic that has no indication that by doing an unrelated action (making the code as asynchronous as need to avoid holding threads) it could actually break the desired semantics.

Reply:
Yes, the current flow in group coordinator is updating the state, queuing up a corresponding log record for commit, and calling the callback when the log record is committed. It needs the log records for a group to be queued up and committed in the same order as the updating of the state so that (1) the latest state can be replayed from the log records correctly in case of failure and (2) the callback can be called in the updating order. Currently, this is achieved through holding the group lock during updating the state and queuing up of the log record (implemented through `ReplicaManager.append`). The expectation is that queuing up of the log record can be done synchronously. Adding async txn validation dependency inside `ReplicaManager.append` breaks that expectation.

To address this issue, we could move the async txn validation dependency out of `ReplicaManager.append`. Alternatively, we could implement another way of synchronously queuing up a log record since `ReplicaManager.append` does more than just queuing up a record.

I am not sure holding the group lock just for updating the state is enough for ensuring the ordering of the log records.

Reply:
I was saying that appendEntries queues up a log record and we could use it there.

I think Artem was not necessarily disagreeing with that idea but rather that a lock should be used at all.

I think we also discussed that the ordering of the log is not the important part but rather committing stale data. It was unclear to me from Jason's comments if this validation is fully necessary due to how we can sometimes write stale data to a transaction when a rebalance happens after records are written but before the transaction is committed.

Basically there are two questions I see:

1. Can we make this code more async friendly to avoid locking? (And does the new group coordinator do this?)
2. Do we really need the lock if we can write inconsistent data outside the typical locking mechanisms and we handle it correctly in those cases?

Reply:
> I was saying that appendEntries queues up a log record and we could use it there.
> 
> I think Artem was not necessarily disagreeing with that idea but rather that a lock should be used at all. My understanding of his comment was that we should move to a model where changes are pending and those state transitions block further work.
> 
> I think we also discussed that the ordering of the log is not the important part but rather committing stale data. It was unclear to me from Jason's comments if this validation is fully necessary due to how we can sometimes write stale data to a transaction when a rebalance happens after records are written but before the transaction is committed.
> 
> Basically there are two questions I see:
> 
> 1. Can we make this code more async friendly to avoid locking? (And does the new group coordinator do this?)
> 2. Do we really need the lock if we can write inconsistent data outside the typical locking mechanisms and we handle it correctly in those cases?

Reply:
@hachikuji Yeah, that's right. I think that we could do it as follow:
1) Apply first verification (e.g. is it the right coordinator, is the member valid, etc)
2) Prepare the transaction (the async call to the transaction coordinator if needed)
3) Re-verify
4) State change + write

This would reduce the likelihood of preparing the transaction for nothing but it could still happen because the state could change in between. I think that this is possible regardless of the model that we use.

Reply:
I agree with @junrao here. 

> The expectation is that queuing up of the log record can be done synchronously.

However, it does not mean that it should be written to the log. The minimum is to guarantee the ordering of the records based on the state changes.

Reply:
> I think we also discussed that the ordering of the log is not the important part but rather committing stale data.

Is that true? It seems that you want the latest log to reflect the latest state, right?

> I was saying that appendEntries queues up a log record and we could use it there.

Yes, the only thing is whether appendEntries can queue up the log record in the right order. I am not sure simply acquiring the group lock when asynchronously calling appendEntries achieves this.

Reply:
My understanding from David is that the main thing we must guarantee is that we can't update the state again until the corresponding records are written to the log. From that, I thought that as long as those steps are synchronous (and I suppose we need a lock for now) that is ok.

The ordering of the requests themselves is not guaranteed already.

Reply:
> The ordering of the requests themselves is not guaranteed already.

It's true that there is not strong ordering guarantee among different clients. However, the ordering how the group coordinator serializes them in the log seems important. 

Consider an example. A group coordinator receives a joinGroup request from client 1 first and generates a group record rec1=GroupA(members: client1). It then receives joinGroup request from client 2 and generates a group record rec2=GroupA(members: client1, client2). If rec1 comes after rec2 in the log and the coordinator fails over, the new coordinator will restore the group state to include only client1 as its members and lose the valid member client2.

Consider another example that involves consumer offsets. A group coordinator starts with a state in which client 1 owns partition 1. It receives an offset commit request from client 1 on partition 1. This is a valid request since client 1 owns partition 1. So, it generates rec3= GroupA(offsetForPartition1=100). A rebalance happens. The coordinator re-assigns partition 1 to client 2 and generates rec4= GroupA(partition1 owned by client 2). If rec3 comes after rec4 in the log and the coordinator fails over, when the new coordinator rebuilds its state, rec3 will look like invalid since it happens when client1 no longer owns partition 1.

Reply:
The serialization can be done asynchronously, without holding locks and relying on synchronous execution.  It looks like it's should be that hard to support that in the new group coordinator https://github.com/apache/kafka/pull/14705.